### ISSUE TITLE
Video Driver Enhancements

### ISSUE THREAD

**Issue by bradallred** [gemrb/gemrb#1771]
This is something I've been working on. Its not ready to show off yet (much of it is unimplemented still), but knowledge of this might be helpful for ongoing BGEE work, in particular the changes for BAMImporter, Font, and the addition of Atlases.

I'll fill this in with more details eventually, but for now here is an almost complete diagram:
![refactor](https://user-images.githubusercontent.com/1126744/211166529-4968b313-aefd-4ec7-999c-055c77cf23d2.png)

Class removals:
TTFFont
TTFFontManager
BAMFont
BAMFontManager
SpriteSheet
GlyphAtlasPage
GlyphIndexEntry
SDLSurfaceSprite2D
SDLTextureSprite2D

Anything for SDL1, SDL20XXX stuff moves to SDLVideo

Class/Interface Additions:
TTFImporter
PaletteManager
BufferAtlas
PixelBuffer
StreamingVideoBuffer
SDLSoftwareRenderer
SDLOpenGLRenderer

## Regarding SDL 1.2 and OpenGL

There is one annoying hitch: maintaining a standard SDL_Renderer backend becomes more annoying and its already of questionable value due to its current limitations. My proposal then, is this: nix the non-OpenGL capabilities for the current SDL2 renderer. This will simplify the code and bugs/maintenance for that renderer a ton. We then create a second SDL2 renderer that use software rendering and promote it as our new reference implementation so we can just wipe SDL 1.2 entirely (the software rendering bits from it will move to the SDL2 software renderer).

This abstraction sets things up nicely for ports to other rendering backends. Adding a new rendering backend would be able to just add the relevant rendering bits and compile them directly into the SDL video driver and set the relevant setting in the cfg instead of the ifdef soup and hacks we are currently running with. 

Likewise this helps with non SDL ports (SMFL, console, whatever) since the video driver now only cares about creating buffers, drawing geometry, and rendering textures. Any port that wants/needs to rely on software rendering can use the non SDL specific software rendering functions we have (including RLE capabilities). Hardware based rendering is reduced to implementing two relatively simple interfaces.

**Comment by bradallred** [comment]
@MarcelHB does any of this seem helpful to what you are doing?

**Comment by MarcelHB** [comment]
I will have a deeper look at some later point. The only thing that comes to my mind: SDL3 was announced for coming up eventually. Maybe take [this doc](https://github.com/libsdl-org/SDL/blob/main/docs/README-migration.md) consideration for forward compatibility/facilitation?

**Comment by bradallred** [comment]
Yes, I am watching SDL3 development. We wont even use `SDL_Renderer` for our SDL 3 implementation. We will use the new `SDL_gpu` API that supports [SDL shaders](https://github.com/libsdl-org/SDL_shader_tools). See #1501.

**Comment by MarcelHB** [comment]
Ok, sounds good to me so at a first glance.

Looking at the first bits of EE, I encountered how our current pipeline and SDL (concerning v2) isn't much helpful w. r. t. the following:

So if a lot of textures are now coming in DXT1/5, any major API for the last 2.5 decades knows how read this on the GPU. For us: no software decompression into ARGB8888, less bandwidth towards the GPU.

Second, the BAM/MOS/TIS data should suit as UV-mapping in the shader. But that also requires uploading multiple rectangular primitives with a second vertex stream. Remembering the hacks around the stencil rendering, this is somewhere between unmaintainable or impossible in conjunction with SDL maintaining its own state in between. 

Currently I don't fully get my head around how this additional data would be carried around in that class diagram, and when and how to do what operation in either render path, and how much of that `SDL_gpu` will be helpful, but first I justed wanted to highlight how the time has come to do more on the GPU than before. :wink:

The only thing I'm chewing on a little is how to keep the `PixelFormatIterator` around in order to do GUI hit tests. This will still require to have textures hanging around in memory, and do at least a partial decompression.

**Comment by bradallred** [comment]
> So if a lot of textures are now coming in DXT1/5, any major API for the last 2.5 decades knows how read this on the GPU. For us: no software decompression into ARGB8888, less bandwidth towards the GPU.

It's a shame SDL 2 doesn't support this. Maybe SDL 3 will. We have yet to see the new GPU API.

> Second, the BAM/MOS/TIS data should suit as UV-mapping in the shader. But that also requires uploading multiple rectangular primitives with a second vertex stream. Remembering the hacks around the stencil rendering, this is somewhere between unmaintainable or impossible in conjunction with SDL maintaining its own state in between.

> Currently I don't fully get my head around how this additional data would be carried around in that class diagram, and when and how to do what operation in either render path, and how much of that SDL_gpu will be helpful, but first I justed wanted to highlight how the time has come to do more on the GPU than before.

I don't think I understand what you are getting at. Our API is simply "take this rectangular portion of this texture and render it to this location on another texture with these settings". How does BGEE differ from that model?

> The only thing I'm chewing on a little is how to keep the PixelFormatIterator around in order to do GUI hit tests. This will still require to have textures hanging around in memory, and do at least a partial decompression.

Sprites are now immutable, so they don't currently keep the data around, and yes, hit testing is broken because of this. My plan was to just use the `Bitmap` class to create an optional hitmap for a `Sprite`. Which is much much smaller than the pixel data.

**Comment by MarcelHB** [comment]
> I don't think I understand what you are getting at. Our API is simply "take this rectangular portion of this texture and render it to this location on another texture with these settings". How does BGEE differ from that model?

Right now, we have `ImageMgr::GetSprite2D()`. For BAMv2, this means that I decompress and copy different tiles from one (and hopefully just one, although the specs don't limit that) PVRZ file into one larger Sprite2D. These tiles appear to be arbitrarily scattered across that DXT texture inside, and the BAM frames' data blocks tell me where to copy from. So far, from what I've seen, `GetSprite2D` always returns spatially coherent data and we treat it like this. And if we want to blit any part (like just one glyph) of it into another texture, or the screenbuffer, there is just one set of a source and destination rectangle description fed into the `SDL_RenderCopy`.

With this new format, there are arbitrarily many rectangles to be copied together _before_ spatial consistency, so that's what I'm doing for now.

But in a best case I don't have this copying, we'd preferably send this DXT texture untouched, as a whole to the GPU and let that do the work. You are right, this can also be done by hardware blitting instructions (OpenGL3+), or analogously calls to `SDL_RenderCopy` -- instead of mapped primitives and UV coords (not sure what's better for this use case). But the point is: In order to avoid software copying for either case, we need the sprite to carry compressed pixels of the DXT _and_ all the tile rectangle data that make this particular sprite rendered, and do many `SDL_RenderCopy` where there has been one, or choose either discussed method for a full acceleration.

**Comment by bradallred** [comment]
Right, but there is no backend support for compressed textures. We are limited to an older GL context, anyway. Even if there were we would still want to have a software renderer.

I know it's not optimal, but there is really no need to complicate things. Just build a decompressed BufferAtlas (the "larger sprite2d") and we keep rendering with a single rect. In the end this is simpler to build and plenty fast even on a mobile.

Maybe we'll be fortunate and SDL3 will have something, but

**Comment by MarcelHB** [comment]
> Maybe we'll be fortunate and SDL3 will have something, but

I think the interesting part of your statement is missing? :wink:

My idea is to challenge how much this class diagram will allow the following setups to be selected at run time:

1. Having a reference renderer (SDL2). `SDL_` calls everywhere, no mixing in OpenGL, and and just doing what's available in the lib, or by hand (texture decompression).
2. Having a fully accelerated renderer: SDL is limited to opening a window and doing a little context setup, but otherwise the lib will only handle input and events anymore. Similar to that OpenGL backend until some years ago, but hopefully less hackish.
3. Having a semi/good faith-accelerated renderer, which is something like the one right now with optional OGL, or maybe an `SDL_gpu`-assisted thing in the future (I've never looked at this so far) for addressing more platforms regardless of API or API version, or one of a common denominator.

So there are implementation classes such that none of these ways will interefere with the other while also not restricting any of the improvement ideas that I mentioned above.

From what I see, there is one little modification required for this in this chart: The importers don't just do `GetSprite2D` but something like `GetRawData2D`. This returned object is defined by the data source, not the video backend, and contains the spectrum of required information:  format, dimension, blob?, blob + palette?, blob tiles? (BAMv2).

With this information, the backend can now decide what's necessary, or what's best for the given `RawData2D`, and return an immutable `Sprite2D` (the class is abstract): The reference renderer will do the full decompression and assembly, the accelerated renderer will just acquire UV-information.

Wouldn't this change enable even more without touching anything else, mostly how GemRB decides to work?

**Comment by MarcelHB** [comment]
A very different side not: Can we get rid of the pixels' `void*` that is still around in a sprite? Like using `std::vector<uint8_t*>` or some shared pointer around this. I just noted that doing `malloc`/`free` and lacking type-defined lifetime semantics is somewhat conflicting my idea of modern C++. :wink: But I did not know whether there is a good reason not do so yet.

**Comment by bradallred** [comment]
> I think the interesting part of your statement is missing?

heh, not really. just expressing skepticism 😉 

The backend is abstracted enough that if you were so inclined you can indeed just use SDL for events and window and write a renderer directly in OpenGL/Vulkan/Metal/DX/Whatever. I'm personally _not_ inclined. I don't see what is to be gained by spending the time writing that much code to only work on a very specific backend. I'd rather spend that time working on an SDL 3 backend that will continue to run on macOS (and many more systems) when Apple finally removes OpenGL.

If you're interested just for the sake of interest, I get that. Go ahead. But if your interest is performance driven, I'd encourage you to simply wait and measure before deciding what needs to be done.

Ok, now lets talk some details that are missing from this diagram:

>  very different side note: Can we get rid of the pixels' void* that is still around in a sprite?

You must have missed my earlier statement. Sprite2D is no longer mutable, and has no pixel data at all. It is just a rectangle and a shared_ptr to a `VideoBuffer` where the rectangle is the area of the `VideoBuffer` to copy from when rendering. So yes, the `pixels` member is gone.

If we were so inclined, we could use a set of rectangles, but I'm still leaning towards just fully decompressing the data when building the `BufferAtlas`.

>  The importers don't just do GetSprite2D but something like GetRawData2D

`Sprite2D` is no longer directly constructible with pixel data anyway. The `Sprite2d` ctor takes a rect and a `shared_ptr<VideoBuffer>`.

You can use a `Video` method to create a `VideoBuffer` from the data then create a `Sprite2D` from that. The main way to get a Sprite2D is from the `AnimationFactory` of which `BamImporter` and `TTFImporter` are subclasses. The other importers wouldn't create `Sprite2D`, they create `PixelBuffers` (with an associated `PixelFormat`) which can be sent to `Video` to create a `VideoBuffer`. You can of course create a `Sprite2d` from that, but you can also just draw using `VideoBuffer`, so you don't really need a sprite unless you want to save rendering options (or for things like a hitmap).

**Comment by MarcelHB** [comment]
> If you're interested just for the sake of interest, I get that. Go ahead. But if your interest is performance driven, I'd encourage you to simply wait and measure before deciding what needs to be done.

I was extrapolating from the RPi performance on the old game data and what was required to achieve both performance and the full feature set. So yes, my concern is performance. Not in the very first steps, but in a way that does not require another turn of major structural refactoring later when actually trying to speed up.
 
> You must have missed my earlier statement. Sprite2D is no longer mutable, and has no pixel data at all. It is just a rectangle and a shared_ptr to a `VideoBuffer` where the rectangle is the area of the `VideoBuffer` to copy from when rendering. So yes, the `pixels` member is gone.

Sorry, my focus was simply on getting rid of any C-style `void*` and this manual allocation (as I had to do for BAMv2) with no indication of lifetime management from the outside, in favor of C++ types that do. Not so much the `pixels` member in particular.

> If we were so inclined, we could use a set of rectangles, but I'm still leaning towards just fully decompressing the data when building the `BufferAtlas`.
> 
> > The importers don't just do GetSprite2D but something like GetRawData2D
> 
> `Sprite2D` is no longer directly constructible with pixel data anyway. The `Sprite2d` ctor takes a rect and a `shared_ptr<VideoBuffer>`.

Ok, then I meant to also consider DXT raw data as being represented by `VideoBuffer`. 

> You can use a `Video` method to create a `VideoBuffer` from the data then create a `Sprite2D` from that. The main way to get a Sprite2D is from the `AnimationFactory` of which `BamImporter` and `TTFImporter` are subclasses. The other importers wouldn't create `Sprite2D`, they create `PixelBuffers` (with an associated `PixelFormat`) which can be sent to `Video` to create a `VideoBuffer`. You can of course create a `Sprite2d` from that, but you can also just draw using `VideoBuffer`, so you don't really need a sprite unless you want to save rendering options (or for things like a hitmap).

Some questions then for me to get a solid understanding (btw. what's the dialect of this UML-diagram? I'm only sure about "-ᐅ" for "being a sub-class of" and "-⨁" as "has"):

1. Let's say we have `PVRZImporter` or `PNGImporter` or any other raw data importer: They return a `PixelBuffer`. This and the associated `PixelFormat`, to what extent does this need to be normalized already? Does this always require something to be strictly like "X bits of R, G, B, A, plus masks, or palette" (as the current `PixelFormat` class is designed like)? Or does this also support symbolic information, like *that's DXT* in order to let `Video` decide what to do best w. r. t. the resulting `VideoBuffer`? If it was symbolic, the `Video` could then decide to do nothing, or to recode the data into something the backend is capable of -- maybe with the help of a suitable `PixelFormatIterator` for input and output. This could move the decision of whether to decompress data or not to a suitable point.

2. Isn't `MOSImporter` and `TISImporter` actually the same kind as a `BAM/TIFImporter`? It's not on the diagram, so I was wondering whether the two are just examples or something different for you.

3. Looking at somthing like `BAMImporter`, or `AnimationFactory` in general, that is supposed to dispense `Sprite2D`s. Can't this ask `Video` to create a suitable `Sprite2D` from all that it knows? Like *I need this `VideoBuffer`, with those rectangles I need here and there, and maybe I also have a `Palette`*. And now `Video` returns a specific implementation of `Sprite2D` for its own needs later on? What that is exactly remains opaque to anyone outside.

**Comment by bradallred** [comment]
> I was extrapolating from the RPi performance on the old game data and what was required to achieve both performance and the full feature set. So yes, my concern is performance.

Then let's wait and measure. The main bottleneck from previous measurements was `Palettes` requiring constant conversion and reupload of pixel data. Now that `Palette`s are backed by a `BufferAtlas` we use a shader to apply palettes which is _much_ faster. Consolidating many `Sprite2D`s into a single texture also helps performance.

> Ok, then I meant to also consider DXT raw data as being represented by VideoBuffer.

Correct, but again, the backend doesn't support DTX and none of the generic cross platform solutions currently do either. You would have to build an entirely new renderer backend. You can still use the SDL2 video driver. I don't think this will be a particularly good use of my effort tho.

The current approach is to decompress/convert to the native 32bit RGBA `PixelBuffer` flavor that `SDL_Renderer` uses and 
build a `VideoBuffer`/`BufferAtlas` from that.

> what's the dialect of this UML-diagram?

It's not UML. I just used a UML tool to create it. The text in the diagram pretty much explains everything.

> Does this always require something to be strictly like "X bits of R, G, B, A, plus masks, or palette" (as the current PixelFormat class is designed like)?

Yes, the `PixelBuffer`s are limited to some form of RGBA. There is also a `Video` method for creating a YUV buffer. There is no reason why we couldn't have other methods that take whatever arbitrary data, but again, there isn't really a point since the backend cant do anything with it other than convert it to RGBA anyway, so we may as well keep things simple and do the conversion on the importer side IMO.

> Isn't MOSImporter and TISImporter actually the same kind as a BAM/TIFImporter? It's not on the diagram, so I was wondering whether the two are just examples or something different for you.

I haven't gotten to that point yet is all. TIS is a tile set, so it too probably should be backed by an atlas, and yes probably an `AnimationFactory`. MOS is just a single image tho isn't it? Probably just have it fetch a `PixelBuffer` then and create a `VideoBuffer` from that.

> Looking at somthing like BAMImporter, or AnimationFactory in general, that is supposed to dispense Sprite2Ds. Can't this ask Video to create a suitable Sprite2D from all that it knows? Like I need this VideoBuffer, with those rectangles I need from here and to there, and maybe I also have a Palette.

Yes, they build a `BufferAtlas`, which is one or more pages of `VideoBuffer` with all the sprites layed out and a set of rectangles for the `Sprite2D` boundaries. Yes, `BufferAtlas` has an optional `Palette` that is assigned to each `Sprite2D` (which can then have their own overriding `Palette`). 

>  Video returns a specific implementation of Sprite2D for its own needs later on? What that is exactly remains opaque to anyone outside.

No, Video knows absolutely nothing about `Sprite2D`. `Video` works exclusively with `VideoBuffer` (or primitives like points/lines/whatever). But yes, the `VideoBuffer` is opaque outside of `Video`. Actually, it's opaque to the VideoDriver too, its only the actual rendering backend that would know anything about what it represents.

**Comment by lynxlynxlynx** [comment]
The tilesets are static, so no need for TIS to be connected to `AnimationFactory`. All the background animations are an extra layer of AreaAnimations.

**Comment by bradallred** [comment]
Right, they are all just atlases that take a key and return a sprite. Animations are just a layer on top of that to make a group of sprites.

**Comment by MarcelHB** [comment]
Ok, then let us know as soon as you get something running. So if there is any bottleneck, adjusting early is probably easier.

**Comment by lynxlynxlynx** [comment]
Ah, we currently do shoehorn TIS tiles into Animations. It's for things like opened/closed doors and other overlays.

**Comment by bradallred** [comment]
#1776 raises a good question about wether our drawing API should be in `float` or 'int` coordinates.

We used `int` because of software rendering and that used to be what SDL drawing functions used. SDL 3 uses floats exclusively now, and 2 has had the option to use floats for a while. Seems like the only value to `int` is the software renderer, which can just cast `float` -> `int`

How much would changing the `Video` API to a float based point complicate things elsewhere? I suppose we just make a `Video::Point` that can be implicitly constructed from `Point` to avoid complications, but that introduces overhead.
